server/src/main/java/org/opensearch/action/search/PitService.java

```diff
@@ -0,0 +1,133 @@
+                        deletePitInfos.add(new DeletePitInfo(false, pitSearchContextIdForNode.getPitId()));
+                    }
+                    groupedListener.onResponse(new DeletePitResponse(deletePitInfos));
+                } else {
+                    try {
+                        final Transport.Connection connection = searchTransportService.getConnection(clusterAlias, node);
+                        searchTransportService.sendFreePITContexts(connection, entry.getValue(), groupedListener);
+                    } catch (Exception e) {
+                        logger.error(() -> new ParameterizedMessage("Delete PITs failed on node [{}]", node.getName()), e);
+                        List<DeletePitInfo> deletePitInfos = new ArrayList<>();
+                        for (PitSearchContextIdForNode pitSearchContextIdForNode : entry.getValue()) {
+                            deletePitInfos.add(new DeletePitInfo(false, pitSearchContextIdForNode.getPitId()));
+                        }
+                        groupedListener.onResponse(new DeletePitResponse(deletePitInfos));
+                    }
+                }
+            }
+        }, listener::onFailure);
+    }
+
+    public GroupedActionListener<DeletePitResponse> getDeletePitGroupedListener(ActionListener<DeletePitResponse> listener, int size) {
+        return new GroupedActionListener<>(new ActionListener<>() {
+            @Override
+            public void onResponse(final Collection<DeletePitResponse> responses) {
+                Map<String, Boolean> pitIdToSucceededMap = new HashMap<>();
+                for (DeletePitResponse response : responses) {
+                    for (DeletePitInfo deletePitInfo : response.getDeletePitResults()) {
+                        if (!pitIdToSucceededMap.containsKey(deletePitInfo.getPitId())) {
+                            pitIdToSucceededMap.put(deletePitInfo.getPitId(), deletePitInfo.isSuccessful());
+                        }
+                        if (!deletePitInfo.isSuccessful()) {
+                            logger.debug(() -> new ParameterizedMessage("Deleting PIT with ID {} failed ", deletePitInfo.getPitId()));
+                            pitIdToSucceededMap.put(deletePitInfo.getPitId(), deletePitInfo.isSuccessful());
+                        }
+                    }
+                }
+                List<DeletePitInfo> deletePitResults = new ArrayList<>();
+                for (Map.Entry<String, Boolean> entry : pitIdToSucceededMap.entrySet()) {
+                    deletePitResults.add(new DeletePitInfo(entry.getValue(), entry.getKey()));
+                }
+                DeletePitResponse deletePitResponse = new DeletePitResponse(deletePitResults);
+                listener.onResponse(deletePitResponse);
+            }
+
+            @Override
+            public void onFailure(final Exception e) {
+                logger.error("Delete PITs failed", e);
```

Comment:
Can we log more info on which PITs?

Reply:
This onFailure is never called in our flow, during connection failures or during delete pit failures, we mark the associated pit ids 'isSucceeded' as false, which gives info that deletion failed.

This will only be executed when there is an unhandled exception and doubt we have context on pits at that point.